docs: adr-003 borrow assets (#123)
umee-network/umee: docs/architecture/ADR-003-borrow-assets.md

type MsgBorrowAsset struct {
  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
  Amount   sdk.Coin       `json:"amount" yaml:"amount"`
}
// MsgRepayAsset - a user wishes to repay assets of a borrowed type
type MsgRepayAsset struct {
  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
  Amount   sdk.Coin       `json:"amount" yaml:"amount"`
}
```
Messages must use denominations only in the allow-list. Collateral is always a uToken denomination, and assets are never uTokens.

Asset borrowing respects the user's borrowing limit, which compares collateral uTokens associated with one base asset, to borrowed assets of a second base asset type. Price oracles must be used to compare the values of base assets (e.g. Atoms:Ether.)

It is necessary that messages be signed by the borrower's account. Thus the method `GetSigners` should return the `Borrower` address for all message types above.

### Module State (Keeper) Additions

Using the `sdk.Coins` built-in type, which combines multiple {Denom,Amount} pairs as a single object, the `umee/x/leverage` module may remember open borrow and collateral positions as follows:

```go

> // pseudocode
> // for each borrower address, combine all open borrow positions into one sdk.Coins object:
> keeper[borrowPrefix + borrowerAddress] = BorrowedCoins
> // additionally user collateral settings are stored as a slice of enabled denoms
> keeper[collateralPrefix + borrowerAddress] = []string{denom1,denom2...}

Author: ```suggestion
// pseudocode
// for each borrower address, combine all open borrow positions into one sdk.Coins object:
borrowPrefix | lengthPrefixed(borrowerAddress) = BorrowedCoins
// additionally borrower collateral settings are stored for enabled denoms
collateralPrefix | lengthPrefixed(borrowerAddress) | tokenDenom = true/false
```

- September 29, 2021: Changed design after review suggestions (@toteki, @alexanderbez, @brentxu)

## Status

Proposed

## Context

One of the base functions of the Umee universal capital facility is to allow users to borrow allowed asset types, using their own uTokens (obtained normally, by depositing assets) as collateral.

## Alternative Approaches

- Borrower deposits uTokens into module account when using as collateral, instead of keeping in user wallet, flagging as collateral-enabled, and restricting transfer. We would not longer have to extend the `x/bank` module, but there would be additional message types.
- Allow amounts of uToken to be specifically marked as collateral, rather than toggling collateral on/off for each asset type. This can be used with or without the other change above, and would allow more fine-grained control of collateral by borrowers.

## Decision

The Cosmos `x/bank` module and the existing `umee/x/leverage` deposit features are prerequisites for these new capabilities.

The flow of events is as follows:
- Borrower already has uTokens in their account
- Borrower marks those uTokens as eligible for use as collateral

> - Borrower deposits uTokens into asset facility as collateral - module remembers collateral position

Author: ```suggestion
```

Note that the exchange rate of Assets:uAssets (e.g. `uatom:u/uatom`) will be a shifting exchange rate that grows with interest - see ADR-001.

In contrast, the exchange rate of original assets:borrowed assets (e.g. `uatom:ether`) can only be determined using a price oracle.

The true borrow exchange rate used in the transaction (e.g. `u/uatom:ether`) is derived from the two above.

Note also that as a consequence of uToken interest, the asset value of uToken collateral increases over time, meaning a user who repays positions in full and redeems collateral uTokens will receive back more base assets than they deposited originally, reducing effective interest.

## Detailed Design

For the purposes of borrowing and repaying assets, as well as depositing and withdrawing uToken collateral, the `umee/x/leverage` module does not mint or burn tokens. It uses its module account to store collateral uTokens, its keeper to remember open positions, and the `x/bank` module to perform all necessary balance checks and token transfers.

### Basic Message Types

> To implement the deposit functionality of the Asset Facility, the two common message types will be:

Author: ```suggestion
To implement the borrow/repay/collateral functionality of the Asset Facility, the four common message types will be:
```

keeper[collateralPrefix + borrowerAddress] = CollateralCoins
keeper[borrowPrefix + borrowerAddress] = BorrowedCoins
```

This will be accomplished by adding new prefixes and helper functions to `x/leverage/types/keys.go`, and using the proper codec to marshal `sdk.Coins` into bytes when storing them as values.

### APIs and Handlers
Both CLI and gRPC must be supported when sending the above message types, and all necessary handlers must be created in order to process and validate them as transactions.

### Testing

Assuming a placeholder token allow-list of at least two elements (e.g. `uumee`,`uatom`), and uTokens existing (e.g. `u/uumee`,`u/uatom`), an integration test can be created in which one user account sends a `MsgBorrowAsset` and a `MsgRepayAsset` of the appropriate token types.

## Open Questions
- See ADR-002 open questions on whitelisting asset types and uniquely identifying ibc/ assets regardless of ibc path.

## Consequences

### Positive
- uTokens used as collateral increase in base asset value in the same way that lend positions do. This counteracts borrow position interest.

### Negative

### Neutral
- Borrow feature relies on allow-list of token types

> - Borrow feature relies on price oracles for base asset types

Author: ```suggestion
- Borrow feature relies on price oracles for base asset types
- Borrow interest will rely on not-yet-implemented dynamic rates
```

> # ADR 003: Borrowing assets using uToken collateral

Reviewer: Remember to add the ADR to the ADR README 👍

## Changelog

- September 27, 2021: Initial Draft (@toteki)

## Status

Proposed

## Context

One of the base functions of the Umee universal capital facility is to allow users to borrow allowed asset types, using their own uTokens (obtained normally, by depositing assets) as collateral.

## Alternative Approaches

It might be worth considering adding message types which do two or more of the steps below at once (e.g. deposit base asset, convert to uToken collateral, and receive loan all at once). The individual steps to need to be possible in isolation, though.

## Decision

The Cosmos `x/bank` module and the existing `umee/x/leverage` deposit features are prerequisites for these new capabilities.

The flow of events is as follows:
- Borrower has already used `umee/x/leverage` module to deposit base assets and receive uTokens
- Borrower deposits uTokens into asset facility as collateral - facility remembers collateral position

> - Borrower requests to borrow assets from capital facility - facility checks request, disburses tokens if acceptable, and remembers borrow position

Reviewer: These two items are technically part of the same "step" right? In other words, the user has some uToken and requests to borrow X amount of some Token using some or all of their uToken.

Reviewer: Is there a check in the state machine that allows borrowing of an asset based on the total reserves loaned?

In compound this is referred to as "collateral factor, ranging from 0 to 1".

Reviewer: Also, when depositing on Compound it shows you the max borrow amount. Are we able to do that here as well?

Author: (Replying as if post-discussion edits already made): Since borrowers no longer need to deposit uTokens when using them as collateral, the first step doesn't happen.

The borrowing limit calculation (which uses collateral factor set by governance) happens when allowing/rejecting borrows and when transferring uTokens (to prohibit undercollateralization). We could display it too by doing the same calculation on the frontend.

- Eventually, borrower pays repayment amount (in full or in part) using borrowed asset denomination

> - Borrower may withdraw collateral uTokens at any time as long as minimum overcollateralization requirement remains satisfied

Reviewer: Do we prohibit the case entirely when it's under collateralized, i.e. message fails, or do we automatically liquidate the position?

Author: Prohibit should work.

- On asset repayment amount = 0 for a given borrow position, facility forgets loan.

> - On uToken collateral amount = 0 for a given collateral position, facility forgets collateral.

Reviewer: I don't understand these two points. Perhaps consider rephrasing.

Author: I'll rephrase it. After the incoming edits, the second line won't exist, and the first will indicate that the `leverage` keeper, which stores all open borrow positions, no longer stores them after they are repaid in full.

> Additionally, the following events occur periodically (at the end of every block):

Reviewer: ```suggestion
Additionally, the following events occur at `EndBlock`:
```

- Repayment amounts for open borrow positions increases by a borrowed-token-specific interest rate
- Open borrow positions are checked to see if they fall below liquidation threshold

> The `umee/x/leverage` module keeper will be responsible for remembering each open borrow or collateral position.

Reviewer: What's the difference between a borrow and collateral position? Or are they the same thing?

Author: Obsolete now - collateral position was "asset facility is holding 5 of Bob's u/uatom" and borrow position is "bob owes 0.04 ether".

If the same user account opens multiple of the same type of position (e.g. borrow) in the same token denomination, the second position simply increases the amount of the first.

Additionally it has been discussed that, rather than an account's specific collateral positions being tied to specific borrow positions, the sum of all borrow and collateral positions related to an account is used to check the account's overcollateralization levels and mark its positions for liquidation.

> Note that the exchange rate of Assets:uAssets (e.g. `uatom:u/uatom`) will be a shifting exchange rate that grows with interest - see ADR-001.

Reviewer: nit: I would link to the ADR...`[ADR-001](...)`.

In contrast, the exchange rate of original assets:borrowed assets (e.g. `uatom:ether`) can only be determined using a price oracle.

The true borrow exchange rate used in the transaction (e.g. `u/uatom:ether`) is derived from the two above.

Note also that as a consequence of uToken interest, the asset value of uToken collateral increases over time, meaning a user who repays positions in full and redeems collateral uTokens will receive back more base assets than they deposited originally, reducing effective interest.

## Detailed Design

> For the purposes of borrowing and repaying assets, as well as depositing and withdrawing uToken collateral, the `umee/x/leverage` module does not mint or burn tokens. It uses its module account to store collateral uTokens, its keeper to remember open positions, and the `x/bank` module to perform all necessary balance checks and token transfers.

Reviewer: In the latter case, i.e. borrowing, tokens are minted and sent to the user and when withdrawing the uTokens are burnt. Not sure if this was intended in this sentence.

Author: (Replying as if post-discussion edits already made): No minting or burning should happen with borrow/repay transactions, as the assets being sent around are all base asset types, not uTokens. Any (loaned) assets the user receives or repays were moved from/to the module account.

### Basic Message Types

To implement the borrow/repay/collateral functionality of the Asset Facility, the four common message types will be:
```go
// MsgDepositCollateral - a user wishes to deposit uToken collateral of one or more types
type MsgDepositCollateral struct {
  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
  Amount   sdk.Coins      `json:"amount" yaml:"amount"`
}
// MsgWithdrawCollateral - a user wishes to withdraw uToken collateral of one or more types
type MsgWithdrawCollateral struct {
  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
  Amount   sdk.Coins      `json:"amount" yaml:"amount"`

> }

Reviewer: Why do we need explicit messages for depositing and withdrawing collateral?

In compound, when a user deposits a Token, they are (optionally) able to enable it as "collateral" via a separate message. Maybe we add a `EnableCollateral` field to `MsgLendAsset` to avoid requiring a separate message?

If enabled, borrowing (up to the borrow limit) is is a one-step message.

In fact, I highly recommend you fiddle with https://app.compound.finance/ on the Rinkeby testnet. Send me your ETH address and I'll send you some ETH tokens.

@brentxu does this make sense?

![image](https://user-images.githubusercontent.com/5265989/135129272-4fd248ef-ebc9-44d1-a284-2142bcfbcb23.png)

Author: (This is the big change that will be made in an upcoming commit)

Author: Change is now made - marking as resolved

- September 29, 2021: Changed design after review suggestions (@toteki, @alexanderbez, @brentxu)

## Status

Proposed

## Context

One of the base functions of the Umee universal capital facility is to allow users to borrow allowed asset types, using their own uTokens (obtained normally, by depositing assets) as collateral.

## Alternative Approaches

- Borrower deposits uTokens into module account when using as collateral, instead of keeping in user wallet, flagging as collateral-enabled, and restricting transfer. We would not longer have to extend the `x/bank` module, but there would be additional message types.
- Allow amounts of uToken to be specifically marked as collateral, rather than toggling collateral on/off for each asset type. This can be used with or without the other change above, and would allow more fine-grained control of collateral by borrowers.

## Decision

The Cosmos `x/bank` module and the existing `umee/x/leverage` deposit features are prerequisites for these new capabilities.

The flow of events is as follows:
- Borrower already has uTokens in their account
- Borrower marks those uTokens as eligible for use as collateral

> - Borrower deposits uTokens into asset facility as collateral - module remembers collateral position

Author: Forgot to remove that line, oops

Note also that as a consequence of uToken interest, the asset value of uToken collateral increases over time, meaning a user who repays positions in full and redeems collateral uTokens will receive back more base assets than they deposited originally, reducing effective interest.

## Detailed Design

For the purposes of borrowing and repaying assets, as well as marking uTokens as collateral, the `umee/x/leverage` module does not mint or burn tokens. It uses its keeper to remember open positions and user collateral settings, and the `x/bank` module to perform all necessary balance checks and token transfers.

Additionally, the `x/bank` module must be extended to prevent users from transferring uTokens when doing so would lower their borrow limit below thir current total borrowed amount.

Because borrow limits weight the value of different token types together, the calculation will require the use of price oracles (and placeholder functions before those are available).

### Basic Message Types

To implement the borrow/repay/collateral functionality of the Asset Facility, the four common message types will be:

> ```go

Reviewer: ```suggestion

```go
```

// MsgSetCollateral - a borrower enables or disables a specific utoken type in their wallet to be used as collateral
type MsgSetCollateral struct {
  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
  Denom    string         `json:"denom" yaml:"denom"`

>   Enable   bool           `json:"enable" yaml:"enable"`

Reviewer: nit

```suggestion
  Enabled   bool           `json:"enable" yaml:"enabled"`
```

Author: Resolved since MsgSetcollateral removed

> // MsgSetCollateral - a borrower enables or disables a specific utoken type in their wallet to be used as collateral
> type MsgSetCollateral struct {
>   Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
>   Denom    string         `json:"denom" yaml:"denom"`
>   Enable   bool           `json:"enable" yaml:"enable"`
> }

Author: ```suggestion
```
@alexanderbez We can remove this message entirely if we add `EnableCollateral` field to the existing `MsgLendAsset` as you said - as long as we support users enabling/disabling collateral on a `MsgLendAsset` with amount = 0 too (i.e. toggling it after the fact).

Should I go ahead and remove this `MsgSetCollateral` message type in favor of doing that?

Reviewer: Yes, remove `MsgSetCollateral` :)

// MsgBorrowAsset - a user wishes to borrow assets of one or more allowed types
type MsgBorrowAsset struct {
  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
  Amount   sdk.Coins      `json:"amount" yaml:"amount"`
}
// MsgRepayAsset - a user wishes to repay assets of one or more borrowed types
type MsgRepayAsset struct {
  Borrower sdk.AccAddress `json:"borrower" yaml:"borrower"`
  Amount   sdk.Coins      `json:"amount" yaml:"amount"`
}
```

> Messages must use only whitelisted denominations. Collateral is always a uToken denomination, and assets are never uTokens.

Reviewer: I personally do not care, but I think some folks might get offended by certain nomenclature, so we should stick with "allow-list" and "deny-list" instead of whitelist/blacklist 👍 

```suggestion
Messages must use denominations only in the allow-list. Collateral is always a uToken denomination, and assets are never uTokens.
```

Asset borrowing respects the user's borrowing limit, which compares collateral uTokens associated with one base asset, to borrowed assets of a second base asset type. Price oracles must be used to compare the values of base assets (e.g. Atoms:Ether.)

_Note: The `Coins` type seen in the `Amount` fields can contain multiple token types. The messages above should fail if even one of the coin types fails or is invalid, rather than partially succeeding._

It is necessary that messages be signed by the borrower's account. Thus the method `GetSigners` should return the `Borrower` address for all message types above.

> ### Keeper additions

Reviewer: ```suggestion
### Keeper Additions
```

Using the `sdk.Coins` built-in type, which combines multiple {Denom,Amount} pairs as a single object, the `umee/x/leverage` keeper may roughly remember open borrow and collateral positions as follows:

```go
// pseudocode
// for each borrower address, combine all open borrow positions into one sdk.Coins object:
keeper[borrowPrefix + borrowerAddress] = BorrowedCoins
// additionally user collateral settings are stored as a slice of enabled denoms

> keeper[collateralPrefix + borrowerAddress] = []string{denom1,denom2...}

Author: This is one way we could store user collateral settings (we need to know the list of uToken denoms they've enabled).

We could also avoid using []string by using an sdk.Coins with an sdk.Coin (amount = 1 or some placeholder value) for each enabled denom.

Reviewer: I would store this as follows, and in general recommend this approach to annotating how storage keys are defined in ADRs:

```
collateralPrefix | lengthPrefixed(lenderAddress) | tokenDenom
```

where `|` denotes concatenation

Reviewer: Let's update both key formats in this ADR 👍
